src/Npgsql/Internal/TypeHandlers/ArrayHandler.cs

```diff
@@ -201,72 +196,91 @@ protected async ValueTask<List<TRequestedElement>> ReadList<TRequestedElement>(N
         internal static class ElementTypeInfo<TElement>
         {
             public static readonly bool IsNonNullable =
-                typeof(TElement).IsValueType && Nullable.GetUnderlyingType(typeof(TElement)) is null;
+                typeof(TElement).IsValueType && default(TElement) != null;
 
             public static readonly Type NullableElementType = IsNonNullable
                 ? typeof(Nullable<>).MakeGenericType(typeof(TElement))
                 : typeof(TElement);
         }
 
-        internal static class ArrayTypeInfo<TArrayOrList>
+        internal abstract class ArrayTypeInfo<TArray>
         {
             // ReSharper disable StaticMemberInGenericType
-            public static readonly bool IsArray;
-            public static readonly bool IsList;
-            public static readonly Type? ElementType;
-
-            public static readonly Func<ArrayHandler, NpgsqlReadBuffer, bool, ValueTask<Array>> ReadArrayFunc = default!;
-            public static readonly Func<ArrayHandler, NpgsqlReadBuffer, bool, ValueTask<TArrayOrList>> ReadListFunc = default!;
+            public static readonly Type? ElementType = typeof(TArray).IsArray ? typeof(TArray).GetElementType() : null;
             // ReSharper restore StaticMemberInGenericType
 
-            public static bool IsArrayOrList => IsArray || IsList;
+            public static bool IsArray => ElementType is not null;
```

Comment:
This (and ElementType) should probably be cached as before, they're on the hot path.

Reply:
Caching a null check? This will get inlined and gone

Reply:
Sorry, ignore.